File: atsa/src/main/java/jp/kaiz/atsassistmod/network/payload/SoundPayloads.java
package jp.kaiz.atsassistmod.network.payload;

import jp.kaiz.atsassistmod.ATSAssistMod;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;

import java.util.ArrayList;
import java.util.List;

/** Server → client sound sequence playback (mirrors PacketPlaySounds / PacketPlaySoundsEntity). */
public final class SoundPayloads {
    private SoundPayloads() {}

    private static Identifier id(String path) {
        return Identifier.fromNamespaceAndPath(ATSAssistMod.MODID, path);
    }

    private static void writeOrders(RegistryFriendlyByteBuf buf, List<String> orders) {
        buf.writeVarInt(orders.size());
        for (String s : orders) {
            buf.writeUtf(s == null ? "" : s);
        }
    }

    private static List<String> readOrders(RegistryFriendlyByteBuf buf) {
        int n = buf.readVarInt();
        List<String> list = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            list.add(buf.readUtf());
        }
        return list;
    }

    /** Plays a sound sequence at one or more positions (anchored to an IFTTT block). */
    public record PlaySoundsAt(List<int[]> positions, List<String> orders, float volume) implements CustomPacketPayload {
        public static final Type<PlaySoundsAt> TYPE = new Type<>(id("play_sounds_at"));
        public static final StreamCodec<RegistryFriendlyByteBuf, PlaySoundsAt> CODEC = StreamCodec.of(
                (buf, p) -> {
                    buf.writeVarInt(p.positions.size());
                    for (int[] pos : p.positions) {
                        buf.writeVarInt(pos[0]);
                        buf.writeVarInt(pos[1]);
                        buf.writeVarInt(pos[2]);
                    }
                    writeOrders(buf, p.orders);
                    buf.writeFloat(p.volume);
                },
                buf -> {
                    int n = buf.readVarInt();
                    List<int[]> positions = new ArrayList<>(n);
                    for (int i = 0; i < n; i++) {
                        positions.add(new int[]{buf.readVarInt(), buf.readVarInt(), buf.readVarInt()});
                    }
                    List<String> orders = readOrders(buf);
                    return new PlaySoundsAt(positions, orders, buf.readFloat());
                });
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }

    /** Plays a sound sequence following an entity. */
    public record PlaySoundsEntity(int entityId, List<String> orders, float volume) implements CustomPacketPayload {
        public static final Type<PlaySoundsEntity> TYPE = new Type<>(id("play_sounds_entity"));
        public static final StreamCodec<RegistryFriendlyByteBuf, PlaySoundsEntity> CODEC = StreamCodec.of(
                (buf, p) -> {
                    buf.writeVarInt(p.entityId);
                    writeOrders(buf, p.orders);
                    buf.writeFloat(p.volume);
                },
                buf -> new PlaySoundsEntity(buf.readVarInt(), readOrders(buf), buf.readFloat()));
        @Override public Type<? extends CustomPacketPayload> type() { return TYPE; }
    }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/network/payload/SoundPayloads.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.network.payload

import jp.kaiz.atsassistmod.ATSAssistMod
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier

/** Server -> client sound sequence playback (mirrors PacketPlaySounds / PacketPlaySoundsEntity). */
object SoundPayloads {
    @JvmStatic
    private fun id(path: String): Identifier = Identifier.fromNamespaceAndPath(ATSAssistMod.MODID, path)

    @JvmStatic
    private fun writeOrders(buf: RegistryFriendlyByteBuf, orders: List<String?>) {
        buf.writeVarInt(orders.size)
        for (order in orders) {
            buf.writeUtf(order ?: "")
        }
    }

    @JvmStatic
    private fun readOrders(buf: RegistryFriendlyByteBuf): List<String> {
        val count = buf.readVarInt()
        val list = ArrayList<String>(count)
        repeat(count) {
            list.add(buf.readUtf())
        }
        return list
    }

    /** Plays a sound sequence at one or more positions (anchored to an IFTTT block). */
    data class PlaySoundsAt(
        val positions: List<IntArray>,
        val orders: List<String>,
        val volume: Float,
    ) : CustomPacketPayload {
        fun positions(): List<IntArray> = positions
        fun orders(): List<String> = orders
        fun volume(): Float = volume

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<PlaySoundsAt> = CustomPacketPayload.Type(id("play_sounds_at"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, PlaySoundsAt> = StreamCodec.of(
                { buf, payload ->
                    buf.writeVarInt(payload.positions.size)
                    for (pos in payload.positions) {
                        buf.writeVarInt(pos[0])
                        buf.writeVarInt(pos[1])
                        buf.writeVarInt(pos[2])
                    }
                    writeOrders(buf, payload.orders)
                    buf.writeFloat(payload.volume)
                },
                { buf ->
                    val count = buf.readVarInt()
                    val positions = ArrayList<IntArray>(count)
                    repeat(count) {
                        positions.add(intArrayOf(buf.readVarInt(), buf.readVarInt(), buf.readVarInt()))
                    }
                    PlaySoundsAt(positions, readOrders(buf), buf.readFloat())
                },
            )
        }
    }

    /** Plays a sound sequence following an entity. */
    data class PlaySoundsEntity(
        val entityId: Int,
        val orders: List<String>,
        val volume: Float,
    ) : CustomPacketPayload {
        fun entityId(): Int = entityId
        fun orders(): List<String> = orders
        fun volume(): Float = volume

        override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

        companion object {
            @JvmField
            val TYPE: CustomPacketPayload.Type<PlaySoundsEntity> = CustomPacketPayload.Type(id("play_sounds_entity"))

            @JvmField
            val CODEC: StreamCodec<RegistryFriendlyByteBuf, PlaySoundsEntity> = StreamCodec.of(
                { buf, payload ->
                    buf.writeVarInt(payload.entityId)
                    writeOrders(buf, payload.orders)
                    buf.writeFloat(payload.volume)
                },
                { buf -> PlaySoundsEntity(buf.readVarInt(), readOrders(buf), buf.readFloat()) },
            )
        }
    }
}
